pkg/utils/utils.go

```diff
@@ -274,30 +276,119 @@ func WriteRequestData(request *http.Request, writer *bytes.Buffer) error {
 // WriteVarIntNum writes a variable-length integer to a buffer
 // integer is converted to fixed size int64
 func WriteVarIntNum(writer *bytes.Buffer, num int) error {
-	err := binary.Write(writer, binary.LittleEndian, int64(num))
-	if err != nil {
-		return fmt.Errorf("failed to write varint number: %w", err)
+	if num < 0 {
+		err := binary.Write(writer, binary.LittleEndian, int64(num))
+		if err != nil {
+			return fmt.Errorf("failed to write negative number: %w", err)
+		}
+
+		return nil
+	}
+
+	if num < 0xFD {
+		err := writer.WriteByte(byte(num))
+		if err != nil {
+			return fmt.Errorf("failed to write single byte varint: %w", err)
+		}
+
+		return nil
+	} else if num <= 0xFFFF {
+		if err := writer.WriteByte(0xFD); err != nil {
+			return fmt.Errorf("failed to write varint prefix: %w", err)
+		}
+
+		numUInt, err := to.UInt16(num)
+		if err != nil {
+			return fmt.Errorf("failed to convert number to uint16: %w", err)
+		}
+
+		err = binary.Write(writer, binary.LittleEndian, numUInt)
+		if err != nil {
+			return fmt.Errorf("failed to write varint number: %w", err)
+		}
+
+		return nil
+
+	} else if num <= 0xFFFFFFFF {
+		if err := writer.WriteByte(0xFE); err != nil {
+			return fmt.Errorf("failed to write varint prefix: %w", err)
+		}
+
+		numUInt, err := to.UInt32(num)
+		if err != nil {
+			return fmt.Errorf("failed to convert number to uint32: %w", err)
+		}
+
+		err = binary.Write(writer, binary.LittleEndian, numUInt)
+		if err != nil {
+			return fmt.Errorf("failed to write varint number: %w", err)
+		}
+
+		return nil
+	} else {
+		if err := writer.WriteByte(0xFF); err != nil {
+			return fmt.Errorf("failed to write varint prefix: %w", err)
+		}
+
+		err := binary.Write(writer, binary.LittleEndian, uint64(num))
+		if err != nil {
+			return fmt.Errorf("failed to write varint number: %w", err)
+		}
+
+		return nil
```

Comment:
all these `if` make me dizzy a bit :)

```suggestion
	if n < 0 {
		if err := binary.Write(w, binary.LittleEndian, int64(n)); err != nil {
			return fmt.Errorf("write negative int64: %w", err)
		}
		return nil
	}

	switch {
	case n < 0xFD:
		return w.WriteByte(byte(n))

	case n <= 0xFFFF:
		if err := w.WriteByte(0xFD); err != nil {
			return fmt.Errorf("write varint prefix 0xFD: %w", err)
		}
		return binary.Write(w, binary.LittleEndian, uint16(n))

	case n <= 0xFFFFFFFF:
		if err := w.WriteByte(0xFE); err != nil {
			return fmt.Errorf("write varint prefix 0xFE: %w", err)
		}
		return binary.Write(w, binary.LittleEndian, uint32(n))

	default:
		if err := w.WriteByte(0xFF); err != nil {
			return fmt.Errorf("write varint prefix 0xFF: %w", err)
		}
		return binary.Write(w, binary.LittleEndian, uint64(n))
	}
```

maybe try this way WDYT ? ... just pls test my suggestion if u decide to use it ...

Reply:
changed to switch case